### PR TITLE
test: btt tests for joinswapPoolAmountOut

### PR DESCRIPTION
**Pull request by 0xteddybear**

### CODE REVIEW

File: test/unit/BPool/BPool_JoinswapPoolAmountOut.t.sol [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+  // Valid scenario
+  uint256 public poolAmountOut = 1e18;
+  uint256 public tokenInWeight = 8e18;
+  uint256 public totalWeight = 10e18;
+  uint256 public tokenInBalance = 300e18;
+  // ((((INIT_POOL_SUPPLY+poolAmountOut)/INIT_POOL_SUPPLY)^(1/(tokenInWeight/totalWeight)))*tokenInBalance-tokenInBalance)/(1-((1-(tokenInWeight/totalWeight))*MIN_FEE))
+  // ((((100+1)/100)^(1/(8/10)))*300-300)/(1-((1-(8/10))*(10^-6)))
+  // 3.754676583174615979425132956656691
+  uint256 public maxTokenIn = 3.754676583181324836e18;
```
Comment (0xteddybear) [inline test/unit/BPool/BPool_JoinswapPoolAmountOut.t.sol:21]: yes, they are different :upside_down_face: I prefer the difference between `BMath` and external calc to be explicit (knowing we're 10^-9 off) to be explicit like this to calling `BMath` directly

File: test/unit/BPool/BPool_JoinswapPoolAmountOut.tree [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+│   └── it should revert
+├── when token amount in exceeds max ratio
+│   └── it should revert 
+├── when calculated token amount in is more than expected
+│   └── it should revert
+└── when preconditions are met
+    ├── it emits LOG_CALL event
+    ├── it sets the reentrancy lock
+    ├── it queries the contracts token in balance
+    ├── it emits LOG_JOIN event for token
```
Comment (wei3erHase) [inline test/unit/BPool/BPool_JoinswapPoolAmountOut.tree:16]:
```suggestion
    ├── it queries token in balance
    ├── it emits LOG_JOIN event for token in
```

File: test/unit/BPool/BPool_JoinswapPoolAmountOut.tree [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    ├── it mints the pool shares
+    ├── it sends pool shares to caller
+    ├── it calls _pullUnderlying for token
```
Comment (wei3erHase) [inline test/unit/BPool/BPool_JoinswapPoolAmountOut.tree:19]:
```suggestion
    ├── it calls _pullUnderlying for token in
```

**Review by wei3erHase — APPROVED** [review]
good! 👍

**Pull request merged**